config/config_params.go

```diff
@@ -356,45 +356,50 @@ func (config *Config) resolve() (changed bool, err error) {
 	return
 }
 
-func (config *Config) DatastoreConfig() apiconfig.CalicoAPIConfig {
-	// Special case for etcdv2 datastore, where we want to honour established Felix-specific
-	// config mechanisms.
-	if config.DatastoreType == "etcdv2" {
-		// Build a CalicoAPIConfig with the etcd fields filled in from Felix-specific
-		// config.
-		var etcdEndpoints string
-		if len(config.EtcdEndpoints) == 0 {
-			etcdEndpoints = config.EtcdScheme + "://" + config.EtcdAddr
-		} else {
-			etcdEndpoints = strings.Join(config.EtcdEndpoints, ",")
-		}
-		etcdCfg := apiconfig.EtcdConfig{
-			EtcdEndpoints:  etcdEndpoints,
-			EtcdKeyFile:    config.EtcdKeyFile,
-			EtcdCertFile:   config.EtcdCertFile,
-			EtcdCACertFile: config.EtcdCaFile,
-		}
-		return apiconfig.CalicoAPIConfig{
-			Spec: apiconfig.CalicoAPIConfigSpec{
-				DatastoreType: apiconfig.EtcdV2,
-				EtcdConfig:    etcdCfg,
-			},
-		}
-	}
+func (config *Config) setBy(name string, source Source) bool {
+	_, set := config.sourceToRawConfig[source][name]
+	return set
+}
 
-	// Build CalicoAPIConfig from the environment.  This means that any XxxYyy field in
-	// CalicoAPIConfigSpec can be set by a corresponding XXX_YYY or CALICO_XXX_YYY environment
-	// variable, and that the datastore type can be set by a DATASTORE_TYPE or
-	// CALICO_DATASTORE_TYPE variable.  (Except in the etcdv2 case which is handled specially
-	// above.)
+func (config *Config) explicitlySet(name string) bool {
+	return config.setBy(name, ConfigFile) || config.setBy(name, EnvironmentVariable)
```

Comment:
I feel like the content of this method doesn't really do what I'd expect form the name (how is a datastore setting less explicit than an environment variable setting?), how about searching all the config types?
```
for _, rawConfig := range config.sourceToRawConfig {
    if _, set := rawConfig[name] {
        return true
    }
}
return false
```

Reply:
Well,

1. Wouldn't that include the `Default` source, which I definitely don't want to include?

2. For these specific settings here, it's now not possible for them to have datastore settings.  (i.e. those have been removed from the FelixConfiguration object).

For these particular settings, being set by config file or environment are definitely the two conditions that we are looking for.  Perhaps I just need to improve the naming so as to better reflect that?

Reply:
Yes, could equally be solved with naming.  I don't think we store Default config in that map but I could be wrong.